addons/stock/stock.py

```diff
@@ -895,8 +895,10 @@ def test_assigned(self, cr, uid, ids):
                 if all([x.state != 'waiting' for x in pick.move_lines]):
                     return True
             for move in pick.move_lines:
-                if (move.state) == 'waiting':
-                    move.check_assign()
+                if move.state == 'waiting':
+                    if self.pool['stock.move'].check_assign(
+                            cr, uid, [move.id], {'picking_trg_write': False}):
+                        move.refresh()
```

Comment:
The next lines check if the move state == 'assigned', which can be the result of the previous call. Without a refresh, it seems that the state is not updated at that point and the right action is not taken.

The loop that is now prevented by this change masked this problem because the same method is then fired again with a freshly browsed picking.